### ISSUE TITLE
Unable to give `-Wno-error=#warnings` flag to clang

### ISSUE THREAD

**Issue by DoDoENT** [android-ndk/ndk#161]
I've explained the issue in [this StackOverflow question](https://stackoverflow.com/questions/38396721/how-to-properly-escape-hash-in-android-mk). At first I thought I am doing something wrong, but then I found out that this may be an `ndk-build` issue.

For convenience, I am including the original SO question here:

> I am using `-Werror` flag to treat all warnings as errors in my code. However, I would like compile warnings issued from code to still be treated as warnings. A GCC option for this is `-Wno-error=cpp` which works correctly, however clang option is `-Wno-error=#warnings` which I don't know how to pass to `clang` from `Android.mk`.
> 
> I've tried this:

```
ifeq "$(findstring clang,$(NDK_TOOLCHAIN_VERSION))" "clang"
    LOCAL_CFLAGS += -Wno-error=#warnings
endif
```

> and this (as suggested in [this SO answer](http://stackoverflow.com/a/7860705/213057)):

```
ifeq "$(findstring clang,$(NDK_TOOLCHAIN_VERSION))" "clang"
    LOCAL_CFLAGS += -Wno-error=\#warnings
endif
```

> However, both versions treat `#` as beginning of comment and `ndk-build` passes `-Wno-error=` to compiler (i.e. everything after `=` is ignored) - I inspected that with `ndk-build -n`.
> 
> I even tried

```
ifeq "$(findstring clang,$(NDK_TOOLCHAIN_VERSION))" "clang"
    LOCAL_CFLAGS += -Wno-error=\\#warnings
endif
```

> which produces `-Wno-error=\` compiler flag.
> 
> How to properly pass this compiler flag to clang from `Android.mk` file?
> 
> I am using NDK r12b.
> 
> EDIT:
> I even tried this:

```
ifeq "$(findstring clang,$(NDK_TOOLCHAIN_VERSION))" "clang"
    VAR=\#warnings
    $(info VAR is '$(VAR)')
    LOCAL_CFLAGS += -Wno-error=$(VAR)
endif
```

> and `ndk-build` first outputs message `VAR is '#warnings'` yet compiler arguments contain `-Wno-error=`. This is obviously an NDK issue.

**Comment by DanAlbert** [comment]
*slow clap for Make*

Nice find. Agreed that this looks like an `ndk-build` issue. Most likely it's being re-evaluated as a comment when we expand it.

**Comment by DanAlbert** [comment]
Tentatively r14, but if I finish the r13 stuff earlier than I expect I'll bump it up.

**Comment by DoDoENT** [comment]
We've switched our build system to CMake, so this is not relevant for us anymore. However, I guess there are still people using ndk-build script for building their projects, so this should get fixed somewhere in the future.

**Comment by DanAlbert** [comment]
I'll keep it open, but I haven't heard anyone else encounter this, so I'll think of it as lower priority. Thanks for letting me know.

**Comment by DanAlbert** [comment]
Well, none of the things I tried allowed me to preserve a properly escaped # through an eval call, so here's the gross fix: https://android-review.googlesource.com/#/c/platform/ndk/+/500056

We could do better, but not without a much more invasive change to get rid of eval in the recipe generation.

**Comment by DoDoENT** [comment]
Hi @DanAlbert. I've moved on to CMake-based build system more than a year ago, so this issue is no longer relevant to me.

However, updating gradle plugin to be able to use more modern CMake (i.e. 3.9), or even system CMake would be more useful.

**Comment by DanAlbert** [comment]
Yep, you'd mentioned that before. It still seemed like a change worth making since you're probably not the only one that will come across this.

> However, updating gradle plugin to be able to use more modern CMake (i.e. 3.9), or even system CMake would be more useful.

The Studio folks have been working on this, fwiw. @jomof, is there a bug tracking updating CMake? I couldn't find record of one in my email.

**Comment by DoDoENT** [comment]
> The Studio folks have been working on this, fwiw. @jomof, is there a bug tracking updating CMake? I couldn't find record of one in my email.

I don't know if there are any bug trackers about that, however I've found [this old SO question](https://stackoverflow.com/questions/40966472/can-android-studio-use-the-system-cmake) regarding the same issue.

**Comment by DanAlbert** [comment]
If there isn't one we should get one filed just for the sake of visibility. It is being worked on though. @jomof might be able to give some hints as to the timeline.

**Comment by DoDoENT** [comment]
FYI, the NDK works with CMake 3.9 even without cmake toolchain that is shipped with NDK (since CMake 3.8, it has got native support for Android, which got even better with 3.9). Since Android Studio is [practically useless for C++ development](https://issuetracker.google.com/issues/38300746), I am using Qt Creator for writing Android NDK code and, since it requires at least CMake 3.8 (due to its support for server mode), I've configured it with following variables:

```
CMAKE_ANDROID_ARCH_ABI:STRING=arm64-v8a
CMAKE_ANDROID_NDK:STRING=/Users/dodo/android-sdks/ndk-bundle
CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION:STRING=clang
CMAKE_ANDROID_STL_TYPE:STRING=c++_static
CMAKE_SYSTEM_NAME:STRING=Android
CMAKE_SYSTEM_VERSION:STRING=21
```

and it can successfully build my project (although not running it because QtCreator knows nothing about Java and APK packaging (however, it may be possible to configure it do do so - I haven't explored that out yet)), without using the NDK provided toolchain at all.

**Comment by DanAlbert** [comment]
Merged into r16
